[gui] smartplaylisteditor: convert spinners etc to regular buttons

phil65 opened the pull request:
Next one...this time the order-by button in Smartplaylisteditor.
this PR adds the same header file to GUIDialogSmartPlaylistEditor.cpp as https://github.com/xbmc/xbmc/pull/10593, so remind me to update the PR which gets pulled last. :)
Could have put all commits into one big PR, so sorry for the additional pinging. 

EDIT: all PRs merged into this one.

EDIT: FInished!
![screenshot031](https://cloud.githubusercontent.com/assets/110931/18980048/afa36dbe-86d0-11e6-8c11-174fc483830d.png)

phil65 committed:
[gui] Always show "Add rule" button in smartplaylisteditor rule list

(cherry picked from commit e6daef4f7674d4ba0fda531e73491231ec0e6060)

phil65 committed:
SmartPlaylistEditor.xml - add missing 'add rule' button (reverted from commit fc87499d3dfd95790be7e2df52b094d98b939d6b)

(cherry picked from commit beec7522e84e799e4021d985b88874a38a937ffe)

phil65 committed:
[estuary] change smartplaylistrule button type to "button"

(cherry picked from commit 3d7fb0ee97fb3d00985fb1a5cb4b8a6c756aae4c)

phil65 committed:
[gui] change rule type selection from spinner to selectdialog

(cherry picked from commit 37babb04e5de2dad84379e3f57142c13cf78cf74)

phil65 committed:
[gui] change group-by selection from spinner to selectdialog

phil65 committed:
[estuary] change button type for group-by to "button"

(cherry picked from commit a3ad5a39df941289db3c9d54fe6d1bc15a03d6b9)

phil65 committed:
[gui] smartplaylisteditor: change another two buttons to "regular" button controls

phil65 commented:
Added another commit which converts two more buttons.
If it helps testing or whatever I could also merge the other PR´s into this one.

MilhouseVH commented:
Merge them and I'll include them in test builds - keeping them separate makes that a bit trickier (as they're both patching the same code).

phil65 commented:
@MilhouseVH done.

MilhouseVH commented:
Thanks, will include in the next build.

phil65 commented:
...aaand another commit.

phil65 committed:
[gui] change button type for playlist type selection to "button" + some refactoring

phil65 committed:
[estuary] - adjust smartplaylisteditor layout

phil65 committed:
[gui] change playlist limit selection to "button"

phil65 committed:
[gui] smartplaylisteditor: change edit control to "button"

phil65 committed:
[estuary] change button type for order-by to "button"

phil65 committed:
[gui] change order-by selection from spinner to selectdialog

phil65 committed:
[gui] smartplaylisteditor: change operator button type to "button"

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp:
```diff
@@ -208,14 +207,17 @@ void CGUIDialogSmartPlaylistEditor::OnPopupMenu(int item)
+{ 
+  if (item == static_cast<int>(m_playlist.m_ruleCombination.m_rules.size()))
+    OnRuleAdd();
+  else if (item < 0 || item > static_cast<int>(m_playlist.m_ruleCombination.m_rules.size()))
```
Please move this condition to the top. IMO early returns should always be at the top if there are multiple conditions. Furthermore after an early return you don't need an `else`.

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp:
```diff
@@ -208,14 +207,17 @@ void CGUIDialogSmartPlaylistEditor::OnPopupMenu(int item)
+  else
+  {
+    CSmartPlaylistRule rule = *std::static_pointer_cast<CSmartPlaylistRule>(m_playlist.m_ruleCombination.m_rules[item]);
+    if (CGUIDialogSmartPlaylistRule::EditRule(rule,m_playlist.GetType()))
```
Space after `rule,`

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp:
```diff
@@ -270,33 +272,80 @@ void CGUIDialogSmartPlaylistEditor::OnCancel()
-  CGUIMessage msg(GUI_MSG_ITEM_SELECTED, GetID(), CONTROL_LIMIT);
-  OnMessage(msg);
-  m_playlist.m_limit = msg.GetParam1();
+  const int limits[] = {0, 10, 25, 50, 100, 250, 500, 1000, -1 };
```
You can use an `std::vector` to get rid of the awkward `-1` special item and then you can use a ranged `for` loop.

Paxxi replied:
Ranged for loop works for plain array as well, the requirements are that std:begin and end exists and those are implemented by default

Montellese replied:
@Paxxi: correct but the `-1` would have to go in that case as well which is my main point/concern.

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp:
```diff
@@ -270,33 +272,80 @@ void CGUIDialogSmartPlaylistEditor::OnCancel()
+void CGUIDialogSmartPlaylistEditor::OnName()
+{
+  std::string name = m_playlist.m_playlistName;
+  if (CGUIKeyboardFactory::ShowAndGetInput(name, CVariant{g_localizeStrings.Get(16012)}, false))
```
You can use `CVariant{ 16012 }` directly. The `CGUIKeyboardFactory::ShowAndGetInput()` knows how to deal with that.

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp:
```diff
@@ -546,6 +548,37 @@ void CGUIDialogSmartPlaylistEditor::HighlightItem(int item)
   OnMessage(msg);
 }
 
+std::vector<CGUIDialogSmartPlaylistEditor::PLAYLIST_TYPE> CGUIDialogSmartPlaylistEditor::GetAllowedTypes(std::string mode)
```
`const std::string& mode`

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistRule.cpp:
```diff
@@ -371,54 +371,15 @@ void CGUIDialogSmartPlaylistRule::OnBrowse()
 }
 
-void CGUIDialogSmartPlaylistRule::UpdateButtons()
+std::vector<std::pair<std::string, int>> CGUIDialogSmartPlaylistRule::GetValidOperators(CSmartPlaylistRule rule)
```
`const CSmartPlaylistRule& rule`

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp:
```diff
@@ -270,33 +272,80 @@ void CGUIDialogSmartPlaylistEditor::OnCancel()
-  OnMessage(msg);
-  m_playlist.SetType(ConvertType((PLAYLIST_TYPE)msg.GetParam1()));
+  std::vector<PLAYLIST_TYPE> allowedTypes = GetAllowedTypes(m_mode);
+  CGUIDialogSelect* dialog = (CGUIDialogSelect*)g_windowManager.GetWindow(WINDOW_DIALOG_SELECT);
```
Please use `static_cast`.

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp:
```diff
@@ -270,33 +272,80 @@ void CGUIDialogSmartPlaylistEditor::OnCancel()
-  OnMessage(msg);
-  m_playlist.m_orderField = (SortBy)msg.GetParam1();
+  std::vector<SortBy> orders = CSmartPlaylistRule::GetOrders(m_playlist.GetType());
+  CGUIDialogSelect* dialog = (CGUIDialogSelect*)g_windowManager.GetWindow(WINDOW_DIALOG_SELECT);
```
`static_cast`

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp:
```diff
@@ -311,11 +360,22 @@ void CGUIDialogSmartPlaylistEditor::OnOrderDirection()
-  m_playlist.SetGroup(CSmartPlaylistRule::TranslateGroup((Field)msg.GetParam1()));
+  std::vector<Field> groups = CSmartPlaylistRule::GetGroups(m_playlist.GetType());
+  Field currentGroup = CSmartPlaylistRule::TranslateGroup(m_playlist.GetGroup().c_str());
+  CGUIDialogSelect* dialog = (CGUIDialogSelect*)g_windowManager.GetWindow(WINDOW_DIALOG_SELECT);
```
`static_cast`

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp:
```diff
@@ -345,50 +400,50 @@ void CGUIDialogSmartPlaylistEditor::UpdateButtons()
     SET_CONTROL_LABEL2(CONTROL_NAME, m_playlist.m_playlistName);
   
   UpdateRuleControlButtons();
+  
```
Unneeded newline

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp:
```diff
@@ -345,50 +400,50 @@ void CGUIDialogSmartPlaylistEditor::UpdateButtons()
+  
 
+  if (m_playlist.m_ruleCombination.GetType() == CSmartPlaylistRuleCombination::CombinationOr)
+    SET_CONTROL_LABEL2(CONTROL_MATCH, g_localizeStrings.Get(21426));
```
Could you add comments to these two `SET_CONTROL_LABEL2` lines to know what the new labels are?

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp:
```diff
@@ -420,73 +475,20 @@ void CGUIDialogSmartPlaylistEditor::UpdateRuleControlButtons()
   bool allowed = false;
-  for (unsigned int i = 0; i < allowedTypes.size(); i++)
-    if (type == allowedTypes[i])
+  for (auto allowedType: allowedTypes)
```
Please use curly braces (even though the original code was missing them as well).

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistRule.cpp:
```diff
@@ -465,15 +426,83 @@ void CGUIDialogSmartPlaylistRule::UpdateButtons()
+
+void CGUIDialogSmartPlaylistRule::OnField()
+{
+
```
Unneeded newline

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistRule.cpp:
```diff
@@ -465,15 +426,83 @@ void CGUIDialogSmartPlaylistRule::UpdateButtons()
+void CGUIDialogSmartPlaylistRule::OnField()
+{
+
+  std::vector<Field> fields = CSmartPlaylistRule::GetFields(m_type);
```
You could use `const auto` here.

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistRule.cpp:
```diff
@@ -465,15 +426,83 @@ void CGUIDialogSmartPlaylistRule::UpdateButtons()
+{
+
+  std::vector<Field> fields = CSmartPlaylistRule::GetFields(m_type);
+  CGUIDialogSelect* dialog = (CGUIDialogSelect*)g_windowManager.GetWindow(WINDOW_DIALOG_SELECT);
```
`static_cast`

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistRule.cpp:
```diff
@@ -465,15 +426,83 @@ void CGUIDialogSmartPlaylistRule::UpdateButtons()
+  dialog->Reset();
+  dialog->SetHeading(CVariant{20427});
+  int selected = -1;
+  for (unsigned int i = 0; i < fields.size(); i++)
```
ranged `for` loop.

phil65 replied:
i need `i`... and I would need sth like `for count, element in enumerate(fields)` How do I do that in Cpp? :)

Montellese replied:
Right it doesn't work with ranged `for` loop. But if you use iterators there's a method called `std::distance()` and you use it like `size_t pos = std::distance(fields.begin(), field);` where `field` is the iterator.

phil65 replied:
does that really make the code easier in this case? :)

Montellese replied:
It makes it more generic in case something around it changes.

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistRule.cpp:
```diff
@@ -465,15 +426,83 @@ void CGUIDialogSmartPlaylistRule::UpdateButtons()
+  if (!isValid)
+    m_rule.m_operator = (CDatabaseQueryRule::SEARCH_OPERATOR)std::get<1>(validOperators[0]);
+  m_rule.SetParameter("");
+  UpdateButtons();
```
Please throw in some empty lines in this code block (e.g. before an `if` statement and after the early return). It's hard to read if it's so dense.

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistRule.cpp:
```diff
@@ -465,15 +426,83 @@ void CGUIDialogSmartPlaylistRule::UpdateButtons()
+
+void CGUIDialogSmartPlaylistRule::OnOperator()
+{
+
```
Unneeded newline

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistRule.cpp:
```diff
@@ -465,15 +426,83 @@ void CGUIDialogSmartPlaylistRule::UpdateButtons()
+{
+
+  std::vector< std::pair<std::string, int> > labels = GetValidOperators(m_rule);
+  CGUIDialogSelect* dialog = (CGUIDialogSelect*)g_windowManager.GetWindow(WINDOW_DIALOG_SELECT);
```
`static_cast`

Montellese commented on xbmc/dialogs/GUIDialogSmartPlaylistRule.cpp:
```diff
@@ -465,15 +426,83 @@ void CGUIDialogSmartPlaylistRule::UpdateButtons()
+void CGUIDialogSmartPlaylistRule::OnOperator()
+{
+
+  std::vector< std::pair<std::string, int> > labels = GetValidOperators(m_rule);
```
`const auto`

Montellese reviewed (changes requested):
Looks good apart from the minors. This will require all skins to adjust the dialog right?

phil65 commented:
@Montellese 
addressed all of your comments in a separate commit, except for the loops, see question above.

Yes, this needs some trivial changes for all skins.

phil65 commented on xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp:
```diff
@@ -546,6 +549,37 @@ void CGUIDialogSmartPlaylistEditor::HighlightItem(int item)
 }
 
+std::vector<CGUIDialogSmartPlaylistEditor::PLAYLIST_TYPE> CGUIDialogSmartPlaylistEditor::GetAllowedTypes(const std::string& mode)
+{
```
Should I move this method somewhere else? The method is static and perhaps better belongs to CSmartPlaylist?

phil65 committed:
[gui] SmartPlaylistEditor: some refactoring and code improvements

phil65 commented:
@Montellese switched to using iterators. Anything else?
Thx for reviewing.

phil65 commented:
jenkins build this please

phil65 commented:
@Montellese 	general skinner´s sentiment was to merge this, even though we´re already in a later beta, since needed skin changes are very trivial and our skinners are quite good in keeping up with these changes.	Are you fine with this?

Montellese commented:
Fine with me but don't forget to squash the cleanup commit(s).

phil65 commented:
All follow-up stuff squashed into one commit.
jenkins build this please.

jingai commented:
Why was the Order button (id=19 in SmartPlaylistEditor.xml) changed from togglebutton to button?  It only toggles between two values.
